scripts/prepare_sector_network.py

```diff
@@ -1893,83 +2134,113 @@ def add_storage_and_grids(
     )
 
-    if options["methanation"]:
-        n.add(
-            "Link",
-            spatial.nodes,
-            suffix=" Sabatier",
-            bus0=nodes + " H2",
-            bus1=spatial.gas.nodes,
-            bus2=spatial.co2.nodes,
-            p_nom_extendable=True,
-            carrier="Sabatier",
-            p_min_pu=options["min_part_load_methanation"],
-            efficiency=costs.at["methanation", "efficiency"],
-            efficiency2=-costs.at["methanation", "efficiency"]
-            * costs.at["gas", "CO2 intensity"],
-            capital_cost=costs.at["methanation", "capital_cost"]
-            * costs.at["methanation", "efficiency"],  # costs given per kW_gas
-            lifetime=costs.at["methanation", "lifetime"],
-        )
 
-    if options["coal_cc"]:
-        n.add(
-            "Link",
-            spatial.nodes,
-            suffix=" coal CC",
-            bus0=spatial.coal.nodes,
-            bus1=spatial.nodes,
-            bus2="co2 atmosphere",
-            bus3=spatial.co2.nodes,
-            marginal_cost=costs.at["coal", "efficiency"]
-            * costs.at["coal", "VOM"],  # NB: VOM is per MWel
-            capital_cost=costs.at["coal", "efficiency"]
-            * costs.at["coal", "capital_cost"]
-            + costs.at["biomass CHP capture", "capital_cost"]
-            * costs.at["coal", "CO2 intensity"],  # NB: fixed cost is per MWel
-            p_nom_extendable=True,
-            carrier="coal",
-            efficiency=costs.at["coal", "efficiency"],
-            efficiency2=costs.at["coal", "CO2 intensity"]
-            * (1 - costs.at["biomass CHP capture", "capture_rate"]),
-            efficiency3=costs.at["coal", "CO2 intensity"]
-            * costs.at["biomass CHP capture", "capture_rate"],
-            lifetime=costs.at["coal", "lifetime"],
-        )
+def add_gas_and_h2_infrastructure(
+    n,
+    costs,
+    pop_layout,
+    h2_cavern_file,
+    cavern_types,
+    clustered_gas_network_file,
+    gas_input_nodes_file,
+    spatial,
+    options,
+    logger,
+):
+    """
+    Add storage and grid infrastructure to the network for and gas and hydrogen.
 
-    if options["SMR_cc"]:
-        n.add(
-            "Link",
-            spatial.nodes,
-            suffix=" SMR CC",
-            bus0=spatial.gas.nodes,
-            bus1=nodes + " H2",
-            bus2="co2 atmosphere",
-            bus3=spatial.co2.nodes,
-            p_nom_extendable=True,
-            carrier="SMR CC",
-            efficiency=costs.at["SMR CC", "efficiency"],
-            efficiency2=costs.at["gas", "CO2 intensity"] * (1 - options["cc_fraction"]),
-            efficiency3=costs.at["gas", "CO2 intensity"] * options["cc_fraction"],
-            capital_cost=costs.at["SMR CC", "capital_cost"],
-            lifetime=costs.at["SMR CC", "lifetime"],
+    Parameters
+    ----------
+    n : pypsa.Network
+        The PyPSA network container object
+    costs : pd.DataFrame
+        Technology cost assumptions
```

Comment:
We might want to stick to the convention as

```suggestion
        Cost assumptions for different technologies. Must include gas and hydrogen assumptions.
```

If so, need to be adapted everywhere.